pkg/container/device.go

```diff
@@ -0,0 +1,173 @@
+
+		// If we are privileged, we have access to devices on the host.
+		// If the requested container path already exists on the host, the container won't see the expected host path.
+		// Therefore, we must error out if the container path already exists
+		if c.Privileged() && device.ContainerPath != device.HostPath {
+			// we expect this to not exist
+			_, err := os.Stat(device.ContainerPath)
+			if err == nil {
+				return errors.Errorf("privileged container was configured with a device container path that already exists on the host.")
+			}
+			if !os.IsNotExist(err) {
+				return errors.Wrap(err, "error checking if container path exists on host")
+			}
+		}
+
+		path, err := ResolveSymbolicLink(device.HostPath, "/")
+		if err != nil {
+			return err
+		}
+		dev, err := devices.DeviceFromPath(path, device.Permissions)
+		// if there was no error, return the device
+		if err == nil {
+			rd := rspec.LinuxDevice{
+				Path:  device.ContainerPath,
+				Type:  string(dev.Type),
+				Major: dev.Major,
+				Minor: dev.Minor,
+				UID:   &dev.Uid,
+				GID:   &dev.Gid,
+			}
+			c.Spec().AddDevice(rd)
+			sp.Linux.Resources.Devices = append(sp.Linux.Resources.Devices, rspec.LinuxDeviceCgroup{
+				Allow:  true,
+				Type:   string(dev.Type),
+				Major:  &dev.Major,
+				Minor:  &dev.Minor,
+				Access: dev.Permissions,
+			})
+			continue
+		}
+		// if the device is not a device node
+		// try to see if it's a directory holding many devices
+		if err == devices.ErrNotADevice {
+			// check if it is a directory
+			if e := utils.IsDirectory(path); e == nil {
+				// mount the internal devices recursively
+				// nolint: errcheck
+				filepath.Walk(path, func(dpath string, f os.FileInfo, e error) error {
+					// filepath.Walk failed, skip
+					if e != nil {
+						return nil
+					}
+					childDevice, e := devices.DeviceFromPath(dpath, device.Permissions)
+					if e != nil {
+						// ignore the device
+						return nil
+					}
+					cPath := strings.Replace(dpath, path, device.ContainerPath, 1)
+					rd := rspec.LinuxDevice{
+						Path:  cPath,
+						Type:  string(childDevice.Type),
+						Major: childDevice.Major,
+						Minor: childDevice.Minor,
+						UID:   &childDevice.Uid,
+						GID:   &childDevice.Gid,
+					}
+					c.Spec().AddDevice(rd)
+					sp.Linux.Resources.Devices = append(sp.Linux.Resources.Devices, rspec.LinuxDeviceCgroup{
+						Allow:  true,
+						Type:   string(childDevice.Type),
+						Major:  &childDevice.Major,
+						Minor:  &childDevice.Minor,
+						Access: childDevice.Permissions,
+					})
+
+					return nil
+				})
+			}
+		}
+	}
+	return nil
+}
+
+// ResolveSymbolicLink resolves a possbile symlink path. If the path is a symlink, returns resolved
+// path; if not, returns the original path.
+func ResolveSymbolicLink(path, scope string) (string, error) {
+	info, err := os.Lstat(path)
+	if err != nil {
+		return "", err
+	}
+	if info.Mode()&os.ModeSymlink != os.ModeSymlink {
+		return path, nil
+	}
```

Comment:
why do we need the extra logic here instead of using directly SecureJoin?  Don't we want to solve symlinks that are not the last path component?

Reply:
I am not sure! This was basically copied over from server/. I will see what dropping it does

Reply:
we do need resolveSymbolicLinks for the remaining case in container_create_linux, as securejoin doesn't differentiate between a symlink and a path not existing. In that case, we need to create if it doesn't exist, so we still need the function. for the device case, I think we can drop it